numba/targets/arraymath.py

```diff
@@ -1020,6 +1020,51 @@ def np_triu_impl_multi(m, k=0):
     else:
         return np_triu_impl_multi
 
+def _prepare_array(arr):
+    pass
+
+@overload(_prepare_array)
+def _prepare_array_impl(arr):
+    if isinstance(arr, (types.Array, types.Sequence, types.Tuple)):
+        return lambda x: _asarray(x).ravel()
+    if arr in (None, types.none):
+        return lambda x: np.array(())
+
+if numpy_version >= (1, 12):  # replicate behaviour of NumPy 1.12 bugfix release
+    @overload(np.ediff1d)
+    def np_ediff1d(ary, to_end=None, to_begin=None):
+
+        if isinstance(ary, types.Array):
+            if isinstance(ary.dtype, types.Boolean):
+                raise TypeError("Numpy does not support case where 'ary' has boolean dtype")
+                # Numpy tries to do this: return ary[1:] - ary[:-1] which results in a
+                # TypeError exception being raised
+
+        def np_ediff1d_impl(ary, to_end=None, to_begin=None):
+            # transform each input into an equivalent 1d array
+            start = _prepare_array(to_begin)
+            mid = _prepare_array(ary)
+            end = _prepare_array(to_end)
+
+            # output array dtype determined by ary dtype, per NumPy
+            out_dtype = mid.dtype
```

Comment:
This delightfully contrived example breaks this assumption and I don't think it can be resolved as the type could only be derived at run time by inspecting the size of `mid`, this creating a type unification issue. I therefore think let's not worry about it :)
```python
import numpy as np
from numba import njit

def foo():
    d1 = [np.complex128(x) for x in range(0)]
    a = np.arange(3)
    return np.ediff1d(d1, a)

print(foo())
print(type(foo()))
print(njit(foo)())
print(type(njit(foo)()))
```

Reply:
Ah, lovely - I will update my comment in the first instance.  :)

Reply:
Comment updated.